src/python/pants/backend/python/tasks/python_binary_create.py

```diff
@@ -128,7 +128,10 @@ def _create_binary(self, binary_tgt, results_dir):
       # Dump everything into the builder's chroot.
       for tgt in source_tgts:
         dump_sources(builder, tgt, self.context.log)
-      dump_requirement_libs(builder, interpreter, req_tgts, self.context.log, binary_tgt.platforms)
+      # We need to ensure that we are resolving for only the current platform if we are
+      # including local python dist targets that have native extensions.
+      platforms = ['current'] if self.tgt_closure_has_native_sources() else binary_tgt.platforms
```

Comment:
this seems like implicit behavior that we'd want to at a minimum warn the user about if the platforms they've defined don't match exactly the current platform?

it would probably appear like a bug to a user who wasn't expecting this if the binary they built ended up with only current platform deps if they've defined e.g. osx + linux as target platforms.